File: C#/045.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace ProjectEuler.Problems
{
	using System.Diagnostics;
	using ProjectEuler.Helper;

	class Problem045 : IProblem
	{

		public void CalculateSolution()
		{
			// Each Hex is Tri, so checking Tri is redundant.
			// Iterating the Hexagonals makes more sense, cause below a limit are less hexagonals than pentagonals
			
			// Equality:
			// P(y) = H(z)
			// 0.5y * (3*y - 1) = z * (2 * z - 1)
			// 1.5y² - 0.5y = 2z² - z						| / 1.5
			// y² - y/3 = 4z²/3 - 2z/3						| + (1/6)² (to complete the square) 
			// y² - y/3 + (1/6)² = 4z²/3 - 2z/3 + (1/6)²		
			// (y - 1/6)² = 4z²/3 - 2z/3 + (1/6)²			|Square root
			// y - 1/6 = SQRT(4z²/3 - 2z/3 + (1/6)²)			| + 1/6
			// y = SQRT(4z²/3 - 2z/3 + 1/36) + 1/6

			var fourthirds = new Fraction(4, 3);
			var twothirds = new Fraction(2, 3);
			var oneoverthirtysix = new Fraction(1, 36);

			var sols = new List<Tuple<int, int, int, long>>();	//T, P, H, number
			for (var z = 1; ; z++)
			{
				var y = Math.Sqrt((fourthirds * z * z - twothirds * z + oneoverthirtysix).ToDouble()) + 1 / 6D;
				if (!(y % 1 < Helper.Methods.Epsilon))
					continue;

				//The xth triangular number == the yth pentagonal number == zth hexagonal number
				var x = 2 * z - 1;
				sols.Add(Tuple.Create(x, (int)y, z, Helper.Methods.GetTriangleNumber(x)));

				if (sols.Count == 3)
					break;
			}

			this.Solution = sols[2].Item4.ToString();
			this.Hint = string.Format("= T({0}) = P({1}) = H({2})", sols[2].Item1, sols[2].Item2, sols[2].Item3);
		}


		public string Solution { get; private set; }

		public string Hint { get; private set; }

	}
}

File: Python/045.py
from itertools import count
from math import sqrt


# Equality:
# T(a) = H(c)
# a*(a+1)/2			=	c(2c-1)
# (aa + a)/2		=	2cc - c
# aa + a 			=	4cc - 2c
# aa + a +1/4-1/4	=   4cc - 2c + 1/4-1/4
# (a+1/2)**2 - 1/4	=	(2c - 1/2)**2 - 1/4
# (a+1/2)**2 		=	(2c - 1/2)**2 
# a + 1/2			=	2c - 1/2
# a = 2c - 1

# This is true for EVERY int c, so checking that is redundant


def pentagonal_index(p):
	n = ((24*p+1)**0.5 + 1) / 6
	return int(n) if n.is_integer() else -1

def h(n):
	return n*(2*n-1)


# Iterating over the Hexagonals makes the most sense, cause below a limit are less hexagonals than pentagonals and triangle nums
found = 0
for c in count(1):
	n = h(c)
	b = pentagonal_index(n)

	if (b != -1):		# H(c) is pentagonal
		a = 2*c - 1
		print((n, a, b, c))
		found += 1

	if (found == 3):
		break
